forge/src/main/java/me/lucko/luckperms/forge/ForgeClassPathAppender.java

```diff
@@ -0,0 +1,135 @@
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+public class ForgeClassPathAppender extends URLClassLoader implements ClassPathAppender {
+    private final PluginLogger logger;
+    private final Set<String> invalidClasses;
+    private final Map<String, ClassLoader> parentLoaders;
+
+    public ForgeClassPathAppender(PluginLogger logger) {
+        super(new URL[0], (ModuleClassLoader) Launcher.class.getClassLoader());
+        this.logger = logger;
+        this.invalidClasses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        this.parentLoaders = getParentLoaders((TransformingClassLoader) Thread.currentThread().getContextClassLoader());
+    }
+
+    @Override
+    protected Class<?> findClass(String name) throws ClassNotFoundException {
+        if (invalidClasses.contains(name)) {
+            throw new ClassNotFoundException(name);
+        }
+
+        invalidClasses.add(name);
+        Class<?> validClass = super.findClass(name);
+        invalidClasses.remove(name);
+        return validClass;
+    }
+
+    @Override
+    public void addJarToClasspath(Path file) {
+        // For debugging purposes
+        // this.logger.info("Adding " + file + " to the classpath");
+
+        try {
+            addURL(file.toUri().toURL());
+        } catch (MalformedURLException ex) {
+            throw new RuntimeException(ex);
+        }
+
+        if (this.parentLoaders == null) {
+            return;
+        }
+
+        try (JarFile jarFile = new JarFile(file.toFile(), false)) {
+            for (Enumeration<JarEntry> enumeration = jarFile.entries(); enumeration.hasMoreElements(); ) {
+                JarEntry jarEntry = enumeration.nextElement();
+                if (jarEntry.isDirectory() || !jarEntry.getName().endsWith(".class")) {
+                    continue;
+                }
+
+                if (jarEntry.getName().endsWith("module-info.class")) {
+                    continue;
+                }
+
+                String name = jarEntry.getName();
+                String packageName = name.substring(0, name.lastIndexOf('/')).replace('/', '.');
+                if (this.parentLoaders.putIfAbsent(packageName, this) == null) {
+                    // For debugging purposes
+                    // this.logger.info("Adding " + packageName + " to the parent loaders");
+                }
+            }
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    @Override
+    public void close() {
+        // no-op
+    }
+
+    @SuppressWarnings("unchecked")
+    private Map<String, ClassLoader> getParentLoaders(TransformingClassLoader classLoader) {
+        try {
+            // sun.misc.Unsafe.theUnsafe
+            Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
```

Comment:
I would strongly prefer to not need to use `Unsafe` for this

Reply:
Unfortunately there isn't another option unless modlauncher is open to allowing mods to dynamically add jars at runtime, I did look into using Forges IModLocator but that requires the implementation to know ALL of the jars that need to be loaded in advanced, From what I can tell LuckPerms doesn't know this until after the configuration has been read but that requires loading configuration jars first, One potential solution is to ship the Forge build of LuckPerms with the dependencies required to read the configuration so that the rest of the required jars are known and can be loaded.

This is what happens with normal reflection:
`java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.util.Map cpw.mods.cl.ModuleClassLoader.parentLoaders accessible: module cpw.mods.securejarhandler does not "opens cpw.mods.cl" to module luckperms`

Reply:
I've figured out how to avoid using `Unsafe` by reviewing how Bukkit, Bungee and Sponge are loaded via the `LoaderBootstrap`.

The only downside I've found so far is that Forge events have to be manually registered due to the ASM that Forge generates not taking the event classes classloader in account which results in ClassNotFoundExceptions on event invocation.

[LXGaming/LuckPerms@forge-loader](https://github.com/LXGaming/LuckPerms/commit/263acdac710909dfcc3c50ed3396ee7ddc1a2fa2)